File: DSA/DSA_450/Cpp/14_DP/059_Matrix_Chain_Multiplication.cpp
/*
Problem: Matrix Chain Multiplication
URL: https://practice.geeksforgeeks.org/problems/matrix-chain-multiplication0303/1

Problem Statement:
Given an array p[] which represents the chain of matrices such that the ith matrix Ai is of dimension p[i-1] x p[i]. Find the minimum number of multiplications needed to multiply the chain.

Sample Input/Output:
Input: p = [40,20,30,10,30]
Output: 26000
*/

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int MCM_Recursive(vector<int>& p, int i, int j) {
        /*
        Recursive approach
        Time Complexity: O(2^n)
        Space Complexity: O(n)
        */
        if (i >= j) return 0;
        
        int minCost = INT_MAX;
        for (int k = i; k < j; k++) {
            int cost = MCM_Recursive(p, i, k) + 
                      MCM_Recursive(p, k + 1, j) + 
                      p[i - 1] * p[k] * p[j];
            minCost = min(minCost, cost);
        }
        
        return minCost;
    }
    
    int MCM_Memo(vector<int>& p, int i, int j, vector<vector<int>>& dp) {
        /*
        Memoization approach
        Time Complexity: O(n^3)
        Space Complexity: O(n^2)
        */
        if (i >= j) return 0;
        if (dp[i][j] != -1) return dp[i][j];
        
        int minCost = INT_MAX;
        for (int k = i; k < j; k++) {
            int cost = MCM_Memo(p, i, k, dp) + 
                      MCM_Memo(p, k + 1, j, dp) + 
                      p[i - 1] * p[k] * p[j];
            minCost = min(minCost, cost);
        }
        
        dp[i][j] = minCost;
        return dp[i][j];
    }
    
    int MCM_Tab(vector<int>& p) {
        /*
        Tabulation approach
        Time Complexity: O(n^3)
        Space Complexity: O(n^2)
        */
        int n = p.size();
        vector<vector<int>> dp(n, vector<int>(n, 0));
        
        for (int length = 2; length < n; length++) {
            for (int i = 1; i < n - length + 1; i++) {
                int j = i + length - 1;
                dp[i][j] = INT_MAX;
                
                for (int k = i; k < j; k++) {
                    int cost = dp[i][k] + dp[k + 1][j] + p[i - 1] * p[k] * p[j];
                    dp[i][j] = min(dp[i][j], cost);
                }
            }
        }
        
        return dp[1][n - 1];
    }
};

void Test_MCM() {
    Solution solution;
    
    vector<int> p = {40, 20, 30, 10, 30};
    
    int result1 = solution.MCM_Recursive(p, 1, p.size() - 1);
    cout << "Recursive: " << result1 << endl;
    
    vector<vector<int>> dp(p.size(), vector<int>(p.size(), -1));
    int result2 = solution.MCM_Memo(p, 1, p.size() - 1, dp);
    cout << "Memo: " << result2 << endl;
    
    cout << "Tab: " << solution.MCM_Tab(p) << endl;
}

int main() {
    Test_MCM();
    return 0;
}

File: DSA/DSA_450/Python/14_DP/059_Matrix_Chain_Multiplication.py
"""
Problem: Matrix Chain Multiplication
URL: https://practice.geeksforgeeks.org/problems/matrix-chain-multiplication0303/1

Problem Statement:
Given an array p[] which represents the chain of matrices such that the ith matrix Ai is of dimension p[i-1] x p[i]. Find the minimum number of multiplications needed to multiply the chain.

Sample Input/Output:
Input: p = [40,20,30,10,30]
Output: 26000
"""


class Solution:
    def MCM_Recursive(self, p: list[int], i: int, j: int) -> int:
        """
        Recursive approach
        Time Complexity: O(2^n)
        Space Complexity: O(n)
        """
        if i >= j:
            return 0
        
        min_cost = float('inf')
        for k in range(i, j):
            cost = (self.MCM_Recursive(p, i, k) + 
                   self.MCM_Recursive(p, k + 1, j) + 
                   p[i - 1] * p[k] * p[j])
            min_cost = min(min_cost, cost)
        
        return min_cost
    
    def MCM_Memo(self, p: list[int], i: int, j: int, dp: list[list[int]]) -> int:
        """
        Memoization approach
        Time Complexity: O(n^3)
        Space Complexity: O(n^2)
        """
        if i >= j:
            return 0
        if dp[i][j] != -1:
            return dp[i][j]
        
        min_cost = float('inf')
        for k in range(i, j):
            cost = (self.MCM_Memo(p, i, k, dp) + 
                   self.MCM_Memo(p, k + 1, j, dp) + 
                   p[i - 1] * p[k] * p[j])
            min_cost = min(min_cost, cost)
        
        dp[i][j] = min_cost
        return dp[i][j]
    
    def MCM_Tab(self, p: list[int]) -> int:
        """
        Tabulation approach
        Time Complexity: O(n^3)
        Space Complexity: O(n^2)
        """
        n = len(p)
        dp = [[0] * n for _ in range(n)]
        
        for length in range(2, n):
            for i in range(1, n - length + 1):
                j = i + length - 1
                dp[i][j] = float('inf')
                
                for k in range(i, j):
                    cost = dp[i][k] + dp[k + 1][j] + p[i - 1] * p[k] * p[j]
                    dp[i][j] = min(dp[i][j], cost)
        
        return dp[1][n - 1]


def Test_MatrixChainMultiplication():
    solution = Solution()
    
    p = [40, 20, 30, 10, 30]
    
    result1 = solution.MCM_Recursive(p, 1, len(p) - 1)
    assert result1 == 26000
    
    dp = [[-1] * len(p) for _ in range(len(p))]
    result2 = solution.MCM_Memo(p, 1, len(p) - 1, dp)
    assert result2 == 26000
    
    result3 = solution.MCM_Tab(p)
    assert result3 == 26000


if __name__ == "__main__":
    Test_MatrixChainMultiplication()
